base/math.jl

```diff
@@ -977,11 +993,17 @@ function modf(x::T) where T<:IEEEFloat
     return (rx, ix)
 end
 
-function ^(x::Float64, y::Float64)
+# @constprop aggressive to help the compiler see the switch between the integer and float
+# variants for callers with constant `y`
+@constprop :aggressive function ^(x::Float64, y::Float64)
     yint = unsafe_trunc(Int, y) # Note, this is actually safe since julia freezes the result
     y == yint && return x^yint
     x<0 && y > -4e18 && throw_exp_domainerror(x) # |y| is small enough that y isn't an integer
     x == 1 && return 1.0
+    return pow_body(x, y)
+end
+
+@inline function pow_body(x::Float64, y::Float64)
```

Comment:
Perhaps worth putting this inside the function, so it is clearer the purpose of it?
```suggestion
    return (@inline function pow_body(x::Float64, y::Float64) ... end)(x, y)
```